src/builder/QuarkBuilder.sol

```diff
@@ -278,6 +279,63 @@ contract QuarkBuilder {
             }
         }
 
+        // Check if need to wrap/unwrap token to cover the transferIntent amount
+        // If the existing balance is not enough to cover the transferIntent amount, wrap/unwrap the counterpart token here
+        uint256 existingBalance =
+            Accounts.getBalanceOnChain(transferIntent.assetSymbol, transferIntent.chainId, chainAccountsList);
+        if (
+            existingBalance < transferIntent.amount
+                && TokenWrapper.hasWrapperContract(transferIntent.chainId, transferIntent.assetSymbol)
+        ) {
+            // If the asset has a wrapper counterpart, wrap/unwrap the token to cover the transferIntent amount
+            string memory counterpartSymbol =
+                TokenWrapper.getWrapperCounterpartSymbol(transferIntent.chainId, transferIntent.assetSymbol);
+            uint256 counterpartBalance =
+                getWrapperCounterpartBalance(transferIntent.assetSymbol, transferIntent.chainId, chainAccountsList);
```

Comment:
Oh after simplifying the flow,  this might not need it anymore. 
I think at this late step,  the bridging logics and wrap/unwrap actions are all in placed. Meaning if we reach this step without hitting errors the final balance on dst chain should work smoothly.  
Thus, we can just simply check if transfer works with existing balance, if not enough,  we are "sure" that by unwrapping what's in counterpart realm should be more than enough to cover the TransferIntent.